File: src/main/java/com/ioannuwu/inline/data/SettingsState.java
package com.ioannuwu.inline.data;

import com.ioannuwu.inline.domain.TextStyle;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Objects;

public class SettingsState implements Serializable {

    public static final SettingsState NONE = new SettingsState();

    public int numberOfWhitespaces = DefaultSettings.NUMBER_OF_WHITESPACES;

    public EffectType effectType = DefaultSettings.EFFECT_TYPE;

    public int maxErrorsPerLine = DefaultSettings.MAX_ERRORS_PER_LINE;

    public FontSettingsState font = DefaultSettings.FONT;

    public TextStyle textStyle = DefaultSettings.TEXT_STYLE;
    public boolean oneGutterMode = DefaultSettings.ONE_GUTTER_MODE;

    public SeverityLevelState error = DefaultSettings.ERROR;
    public SeverityLevelState warning = DefaultSettings.WARNING;
    public SeverityLevelState weakWarning = DefaultSettings.WEAK_WARNING;
    public SeverityLevelState information = DefaultSettings.INFORMATION;
    public SeverityLevelState serverError = DefaultSettings.SERVER_ERROR;
    public SeverityLevelState otherError = DefaultSettings.OTHER_ERROR;

    public String[] ignoreList = DefaultSettings.IGNORE_LIST;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SettingsState that = (SettingsState) o;
        return numberOfWhitespaces == that.numberOfWhitespaces && maxErrorsPerLine == that.maxErrorsPerLine &&
                Objects.equals(error, that.error) && Objects.equals(warning, that.warning) &&
                Objects.equals(weakWarning, that.weakWarning) && Objects.equals(information, that.information) &&
                Objects.equals(serverError, that.serverError) && Objects.equals(otherError, that.otherError) &&
                Arrays.equals(ignoreList, that.ignoreList) && Objects.equals(effectType, that.effectType) &&
                Objects.equals(font, that.font) && textStyle.equals(that.textStyle) && oneGutterMode == that.oneGutterMode;
    }

    @Override
    public int hashCode() {
        int result = Objects.hash(numberOfWhitespaces, maxErrorsPerLine, error, warning, weakWarning,
                information, serverError, otherError, effectType, font, textStyle, oneGutterMode);
        result = 31 * result + Arrays.hashCode(ignoreList);
        return result;
    }

    @Override
    public String toString() {
        return "\nSettingsState{" +
                "\nnumberOfWhitespaces=" + numberOfWhitespaces +
                "\neffectType=" + effectType +
                "\nmaxErrorsPerLine=" + maxErrorsPerLine +
                "\nfont=" + font +
                "\ntextStyle=" + textStyle +
                "\noneGutterMode=" + oneGutterMode +
                "\nerror=" + error +
                "\nwarning=" + warning +
                "\nweakWarning=" + weakWarning +
                "\ninformation=" + information +
                "\nserverError=" + serverError +
                "\notherError=" + otherError +
                "\nignoreList=" + Arrays.toString(ignoreList) +
                '}';
    }
}

File: src/main/kotlin/com/ioannuwu/inline2/settings/data/SettingsState.kt
package com.ioannuwu.inline2.settings.data

import com.ioannuwu.inline.data.*
import java.util.*

class SettingsState {

    var numberOfWhitespaces: Int = DefaultSettings.NUMBER_OF_WHITESPACES

    var effectType: EffectType = DefaultSettings.EFFECT_TYPE

    var maxErrorsPerLine: Int = DefaultSettings.MAX_ERRORS_PER_LINE

    var font: FontSettingsState = DefaultSettings.FONT

    var textStyle: TextStyle = DefaultSettings.TEXT_STYLE
    var oneGutterMode: Boolean = DefaultSettings.ONE_GUTTER_MODE

    var error: SeverityLevelState = DefaultSettings.ERROR
    var warning: SeverityLevelState = DefaultSettings.WARNING
    var weakWarning: SeverityLevelState = DefaultSettings.WEAK_WARNING
    var information: SeverityLevelState = DefaultSettings.INFORMATION
    var serverError: SeverityLevelState = DefaultSettings.SERVER_ERROR
    var otherError: SeverityLevelState = DefaultSettings.OTHER_ERROR

    var ignoreList: Array<String> = DefaultSettings.IGNORE_LIST

    fun copy(): SettingsState {
        val state = SettingsState()
        state.numberOfWhitespaces = numberOfWhitespaces
        state.effectType = effectType
        state.maxErrorsPerLine = maxErrorsPerLine
        state.font = font
        state.textStyle = textStyle
        state.oneGutterMode = oneGutterMode
        state.error = error
        state.warning = warning
        state.weakWarning = weakWarning
        state.information = information
        state.serverError = serverError
        state.otherError = otherError
        state.ignoreList = ignoreList
        return state
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as SettingsState
        return numberOfWhitespaces == that.numberOfWhitespaces && maxErrorsPerLine == that.maxErrorsPerLine
                && error == that.error && warning == that.warning && weakWarning == that.weakWarning
                && information == that.information && serverError == that.serverError
                && otherError == that.otherError
                && ignoreList.contentEquals(that.ignoreList)
                && effectType == that.effectType && font == that.font && textStyle == that.textStyle
                && oneGutterMode == that.oneGutterMode
    }

    override fun hashCode(): Int {
        var result = Objects.hash(
            numberOfWhitespaces, maxErrorsPerLine, error, warning, weakWarning,
            information, serverError, otherError, effectType, font, textStyle, oneGutterMode
        )
        result = 31 * result + ignoreList.contentHashCode()
        return result
    }

    override fun toString(): String = """
        SettingsState{
        numberOfWhitespaces=$numberOfWhitespaces
        effectType=$effectType
        maxErrorsPerLine=$maxErrorsPerLine
        font=$font
        textStyle=$textStyle
        oneGutterMode=$oneGutterMode
        error=$error
        warning=$warning
        weakWarning=$weakWarning
        information=$information
        serverError=$serverError
        otherError=$otherError
        ignoreList=${ignoreList.contentToString()}}
        """.trimIndent()
}